core/src/apps/management/sd_protect.py

```diff
@@ -0,0 +1,157 @@
+from trezor import config, ui, wire
+from trezor.crypto import hmac, random
+from trezor.crypto.hashlib import sha256
+from trezor.messages import SdProtectOperationType
+from trezor.messages.Success import Success
+from trezor.pin import pin_to_int
+from trezor.ui.text import Text
+
+from apps.common.confirm import require_confirm
+from apps.common.request_pin import request_pin_ack, request_pin_and_sd_salt
+from apps.common.sd_salt import (
+    SD_SALT_AUTH_KEY_LEN_BYTES,
+    SD_SALT_AUTH_TAG_LEN_BYTES,
+    SD_SALT_LEN_BYTES,
+    commit_sd_salt,
+    remove_sd_salt,
+    set_sd_salt,
+    stage_sd_salt,
+)
+from apps.common.storage import device, is_initialized
+
+if False:
+    from trezor.messages.SdProtect import SdProtect
+
+
+async def sd_protect(ctx: wire.Context, msg: SdProtect) -> Success:
+    if not is_initialized():
+        raise wire.ProcessError("Device is not initialized")
+
+    if msg.operation == SdProtectOperationType.ENABLE:
+        return await sd_protect_enable(ctx, msg)
+    elif msg.operation == SdProtectOperationType.DISABLE:
+        return await sd_protect_disable(ctx, msg)
+    elif msg.operation == SdProtectOperationType.REFRESH:
+        return await sd_protect_refresh(ctx, msg)
+    else:
+        raise wire.ProcessError("Unknown operation")
+
+
+async def sd_protect_enable(ctx: wire.Context, msg: SdProtect) -> Success:
+    salt_auth_key = device.get_sd_salt_auth_key()
+    if salt_auth_key is not None:
+        raise wire.ProcessError("SD card protection already enabled")
+
+    # Confirm that user wants to proceed with the operation.
+    await require_confirm_sd_protect(ctx, msg)
+
+    # Get the current PIN.
+    if config.has_pin():
+        pin = pin_to_int(await request_pin_ack(ctx, "Enter PIN", config.get_pin_rem()))
+    else:
+        pin = pin_to_int("")
+
+    # Check PIN and write salt.
+    salt = random.bytes(SD_SALT_LEN_BYTES)
+    salt_auth_key = random.bytes(SD_SALT_AUTH_KEY_LEN_BYTES)
+    salt_tag = hmac.new(salt_auth_key, salt, sha256).digest()[
+        :SD_SALT_AUTH_TAG_LEN_BYTES
+    ]
+    try:
+        await set_sd_salt(ctx, salt, salt_tag)
+    except Exception:
+        raise wire.ProcessError("Failed to write to SD card")
+
+    if not config.change_pin(pin, pin, None, salt):
+        try:
+            await remove_sd_salt(ctx)
+        except Exception:
```

Comment:
This is not very readable. Should it catch `SdProtectCancelled` only?

Reply:
We should never catch a generic Exception, it's a bad pattern.

Reply:
The purpose of `remove_sd_salt()` here is to clean up the prepared salt file on the SD card in case the user enters a wrong PIN. The clean-up is not necessary for the correct functioning of the sd-protect feature, it's just a matter of not leaving the file there, when the feature is not enabled. So any kind of failure at this point is acceptable. I see two possible exceptions that could trigger this: `OSError` and `SdProtectCancelled`. I am not sure if there are any others. The reason why I used the generic exception is that even in case there happens to be another type of exception (possibly even one added in the future), I still want `wire.PinInvalid` to be raised, because that is the relevant piece of information which should get to the user.

Reply:
Options:
1. Add a comment explaining what I just wrote.
2. Remove the clean-up since it's not essential.
3. Name the individual exceptions that we are aware of. As I explained I am not in favor of this, because I don't think any exception in this case should surpass wire.PinInvalid.

Reply:
1. seems fine

Reply:
Added a comment here and in the two functions below. https://github.com/trezor/trezor-firmware/pull/526/commits/5a7e2fdb7a4fa87372bbd246eddfb4769b9da8b8